EBI upgrade from Python 2 to Python 3

nelson-edalex opened the pull request:
This is a huge upgrade to Python 3, which includes:
- Migration of all code to be compatible with Python 3.
- Updates to package files for macOS and Windows/Linux.
- Addition of a requirements.txt file for dependency management.
- Various modifications to source files to ensure compatibility and functionality.
- Addition of a migration guide in PYTHON3_MIGRATION.md.
- Backup of the original Engine.py file.
- Updated .gitignore to reflect new files and directories.
- Updated README.md to reflect changes and new instructions.
- Updated package scripts for macOS and Windows/Linux to work with Python 3.
- Updated ebi.properties to ensure compatibility with Python 3.
- Updated equellaclient41.py to ensure compatibility with Python 3.
- Updated package scripts to ensure they work with the new Python 3 environment.
- Updated package metadata files to reflect the new Python version.
- Updated setup.py files for both macOS and Windows/Linux packages to ensure they work with Python

Commit:
This is a huge upgrade to Python 3, which includes:
- Migration of all code to be compatible with Python 3.
- Updates to package files for macOS and Windows/Linux.
- Addition of a requirements.txt file for dependency management.
- Various modifications to source files to ensure compatibility and functionality.
- Addition of a migration guide in PYTHON3_MIGRATION.md.
- Backup of the original Engine.py file.
- Updated .gitignore to reflect new files and directories.
- Updated README.md to reflect changes and new instructions.
- Updated package scripts for macOS and Windows/Linux to work with Python 3.
- Updated ebi.properties to ensure compatibility with Python 3.
- Updated equellaclient41.py to ensure compatibility with Python 3.
- Updated package scripts to ensure they work with the new Python 3 environment.
- Updated package metadata files to reflect the new Python version.
- Updated setup.py files for both macOS and Windows/Linux packages to ensure they work with Python

edalex-ian commented on source/Engine.py.backup:
If this is just a copy of Engine.py from before the upgrade. I'd say dump it - as we should just rely on source control version history to see what's the difference.

nelson-edalex replied:
Hi @edalex-ian, the copy is removed.
Please advise if this is good to go and how can I make a release. :)

Commit:
The following is the commit message for your changes:
- Updated README.md to reflect Python 3 support
- Modified Engine.py files for Linux compatibility with Python 3.10+
- Added pyproject.toml for project configuration
- Removed obsolete backup file source/Engine.py.backup

edalex-ian commented on PYTHON3_MIGRATION.md:
```diff
@@ -0,0 +1,161 @@
+2. **Install wxPython**: Run `pip3 install wxPython` to install the GUI framework
+3. **Test Import**: Verify all modules import successfully with Python 3
+4. **Runtime Testing**: Test the application with actual CSV files and EQUELLA connections
+5. **Update README**: Document Python 3 requirements and setup instructions
```
Has this all been actioned?

If so, we should probably remove it. If not, what's the plan?

nelson-edalex replied:
This has been actioned.  I've remove the complete file.  It's the AI's execution plan and I believe we don't need to keep it in the repo.

edalex-ian commented on PYTHON3_MIGRATION.md:
```diff
@@ -0,0 +1,161 @@
+- [ ] Can process items (test mode)
+- [ ] Can upload attachments
+- [ ] Export functionality works
+- [ ] Logging works correctly
```
Have these steps been tested?

nelson-edalex replied:
Yes it's tested.

edalex-ian commented on README.md:
````diff
@@ -5,20 +5,67 @@ The EBI is written in Python and compiled to a standalone version (i.e. will run
+
+#### Linux (Ubuntu/Debian)
+```bash
+sudo apt-get install python3-wxgtk4.0
````
You need a bit more than that. This assumes that GTK+ 3 is available as well as the python dev tools and gettext. I suggest changing this to:

```suggestion
sudo apt-get install build-essential gettext python3-dev libgtk-3-dev python3-wxgtk4.0
```

Indeed, that's what I needed to get wxPython to install.

edalex-ian commented on README.md:
````diff
@@ -5,20 +5,67 @@ The EBI is written in Python and compiled to a standalone version (i.e. will run
+
+#### Linux (Ubuntu/Debian)
+```bash
+sudo apt-get install python3-wxgtk4.0
````
I had a comment here, but it disappear. So here we go again...

This actually needs several more dependencies. Here's the full list I had to install on a relatively new machine to get this to work:

```suggestion
sudo apt-get install build-essential gettext python3-dev libgtk-3-dev python3-wxgtk4.0
```

nelson-edalex replied:
The dependencies are updated.  `wxPython-4.2.0` already include the wxgtk4.0.  The other apt libraries are added.

edalex-ian commented on package-mac/source/ebi/Engine.py:
```diff
@@ -176,7 +176,15 @@ def getPlatform(self):
+                else:
+                    ebiPlatform += ", Linux " + platform.release()
+            except Exception:
+                ebiPlatform += ", Linux " + platform.release()
```
This change concerns me on two levels:

- Immediate: Why does a file sitting in the mac package folder care about other operating systems.
- Big picture: This Engine.py seems to be a variant of source/Engine.py. Indeed, this Engine.py didn't get some of the import updates. So is it relevant?

nelson-edalex replied:
@edalex-ian I removed this folder.  It's a copy of `source` folder for running the build tool.

edalex-ian commented on README.md:
```diff
@@ -5,20 +5,67 @@ The EBI is written in Python and compiled to a standalone version (i.e. will run
+pip3 install -r requirements.txt
+
+# Run from source
+python3 source/ebi.py
```
Good news. I did get this to work on Ubuntu.

(I do note that the help information is woefully incorrect - still talks of Pearson etc.)

<img width="959" height="1049" alt="2026-01-12_175840" src="https://github.com/user-attachments/assets/a9dbae6a-a3e8-4ae6-88dc-b3fd5904391d" />

edalex-ian commented on package-linux/README.md:
````diff
@@ -0,0 +1,135 @@
+pyinstaller ebi.spec
+```
+
+## Output
+
+The build process creates:
+
+- `dist/ebi.exe` - The standalone executable
+- `ebi.zip` - Distribution package containing the executable and source files
````
When using Option 2 on linux, I simply got a 119M `dist/ebi` which seems to run fine.

BUT, no zip file. :thinking:

edalex-ian commented on pyproject.toml:
```diff
@@ -0,0 +1,51 @@
+license = {text = "Apache-2.0"}
+authors = [
+    {name = "Apereo Foundation", email = ""},
+    {name = "Jim Kurian", email = ""}
```
Who's Jim?

I'd argue that seeing this is a new setup, it's probably not required to provide attribution to individuals. Alternatively, even just drop the `authors` entirely.

edalex-ian commented on requirements.txt:
So you've added both `pyproject.toml` and this `requirements.txt` and now the requirements will need to be kept in sync in both.

Typically these days you'd just go with `pyproject.toml`. If you still wanted a `requirements.txt` you'd avoid the need to manage the deps in both places by using something like the following in your `pyproject.toml`.

```toml
[project]
name = "openequella-ebi"
version = "4.73"
dynamic = ["dependencies", "optional-dependencies"]

[tool.setuptools.dynamic]
dependencies = { file = ["requirements.txt"] }
optional-dependencies.dev = { file = ["requirements-dev.txt"] }
```

edalex-ian commented on pyproject.toml:
```diff
@@ -0,0 +1,51 @@
+
+[project]
+name = "openequella-ebi"
+version = "4.73"
```
This version should totally be bumped as part of this. I'd suggest it should go to version 5.0. Sure feature wise etc we expect it to be the same, but there's been significant changes under the hood.

And seeing this is being merged directly to `master` then we need to make sure we bump it now - not after.

(Although, that could be something to consider: Should PR target `develop`, we then can do a proper release cycle for it with some testing by QA.)

edalex-ian commented on source/Engine.py:
```diff
@@ -160,7 +161,8 @@ def __init__(self, owner, Version, Copyright):
                                                 "history",
-                                                "thumbnail"
+                                                "thumbnail",
+                                                "itembody"
```
Oh! So this also includes some feature changes/enhancements? Not just a Python 3 migration...

edalex-ian commented on source/Engine.py:
```diff
@@ -582,19 +628,19 @@ def unicode_csv_reader(self, utf8_data, encoding, dialect=csv.excel, **kwargs):
-                reader = self.unicode_csv_reader(open(self.csvFilePath, "rbU"), self.encoding)
+                reader = self.unicode_csv_reader(open(self.csvFilePath, "r", encoding=self.encoding, newline=''), self.encoding)
                 for row in reader:
                     self.csvArray.append(row)
```
`unicode_csv_reader` seemed cumbersome and possibly redundant. Chatting with Gemini I came up with this refactor of `load_csv` which sees `unicode_csv_reader` dropped.

---

Here is the refactored code. I have removed the `unicode_csv_reader` helper entirely, as the standard library now handles that complexity, and simplified `loadCSV` to be more linear and readable.

This refactor applies **Standard Conventions** (PEP 8 snake_case) and **Function Rules** (Do one thing, Small).

```python
    def load_csv(self, owner=None):
        """Loads CSV data, handling encoding and defaults."""
        self.csv_array = []

        # 1. Guard Condition: Validate owner path if applicable
        if owner and not owner.txtCSVPath.GetValue():
            return

        # 2. Read Data: Encapsulate boundary conditions (BOM) using utf-8-sig
        # - "Prefer data structures" & "standard conventions"
        if os.path.isfile(self.csv_file_path):
            with open(self.csv_file_path, "r", encoding="utf-8-sig", newline='') as f:
                self.csv_array = list(csv.reader(f))

        # 3. Clean Data: Remove empty trailing rows
        # - "Separate concepts vertically"
        while self.csv_array and not self.csv_array[-1]:
            self.csv_array.pop()

        # 4. Default State: specific business logic for empty files
        if not self.csv_array:
            self.csv_array.append(self.column_headings[:])

```

### Key Changes

* **Declarative Encoding**: Replaced the manual `unicode_csv_reader` and magic strings (`'\ufeff'`) with `encoding='utf-8-sig'`. This simplifies the code and reduces "Needles Complexity".
* **Vertical Separation**: The function is structured into distinct steps (Guard, Read, Clean, Default), making it easy to scan and understand "vertically dense" related code.
* **Variable Naming**: Renamed `csvArray` to `csv_array` and `csvFilePath` to `csv_file_path` to follow Python "standard conventions" and avoid "encodings" (type information in names).

**Next Step:** Would you like me to extract the "Clean Data" (trimming) and "Default State" logic into their own small, private methods to strictly enforce the "Do one thing" rule?

edalex-ian commented on source/Engine.py:
```diff
@@ -636,22 +682,22 @@ def csvParse(self,
         receiptFilename = ""
         if not self.export:
-            if itemIdColumn != -1:
+            if itemIdColumn != -1 or rowErrorColumn != -1:
```
Unexpected change.

(That is, has nothing to do with python3 and I can't see why it's needed now.)

edalex-ian commented on source/Engine.py:
```diff
@@ -867,7 +911,8 @@ def csvParse(self,
                                                                         ownerColumn,
-                                                                        collaboratorsColumn)
+                                                                        collaboratorsColumn,
+                                                                        rowErrorColumn)
```
Unexpected change. New feature?

edalex-ian commented on source/Engine.py:
```diff
@@ -988,6 +1058,16 @@ def csvParse(self,
+                writer.writerows(self.csvArray)
+                f.flush()
+                os.fsync(f.fileno())
+            self.echo(time.strftime("%H:%M:%S: ", time.localtime(time.time())) + "CSV file updated successfully")
```
A few things jump out:

1. This function was already 400 lines long!! Extending it like this is a bad idea.
2. This also changes what this function does, if you want to do this kind of thing arguably it should be done in the calling function after this process is done.
3. This is another feature change by the looks of things. :thinking:

edalex-ian commented on source/Engine.py:
```diff
@@ -1331,20 +1408,39 @@ def unzipAttachment():
+                            # If no exact match found, use first result
+                            if not foundExactMatch:
+                                collaboratorID = matchingUserNodes[0].getElementsByTagName("uuid")[0].firstChild.nodeValue
+                                collaboratorIDs.append(collaboratorID)
```
Should be in its own function - no inlined.

edalex-ian commented on source/ebi.properties:
```diff
@@ -1,6 +1,16 @@
 
 [State]
 mainframesize = (1012, 575)
+settingsfile = /Users/nelson/gitlab/consulting/openEQUELLA-ebi/source/test.ebi
```
I'm thinking this line should be removed. Or at least not tied to your machine's setup.

edalex-ian commented on source/MainFrame.py:
```diff
@@ -1081,18 +1080,18 @@ def unicode_csv_reader(self, utf8_data, encoding, dialect=csv.excel, **kwargs):
-            yield [cell.decode(encoding) for cell in row]
+            yield row  # Python 3: csv.reader already returns strings
     
     def LoadCSV(self):
```
As per my comment in `Engine.py` this can no be simplified and `unicode_csv_reader` removed.

edalex-ian commented on source/equellaclient41.py:
```diff
@@ -220,7 +222,16 @@ def _call (self, name, args, returns=1, facade=SOAP_ENDPOINT_V41, ns='http://soa
+            if name == 'getContributableCollections':
+                print(f"\n=== Cookie Debug for {name} ===")
+                print(f"Number of cookies: {len(self._cookieJar)}")
+                print(f"Cookies being sent: {self._cookieJar[:3] if len(self._cookieJar) > 3 else self._cookieJar}")
```
Security concern: Revealing security cookies in logs/terminals etc.

edalex-ian commented on source/equellaclient41.py:
```diff
@@ -247,18 +258,33 @@ def _call (self, name, args, returns=1, facade=SOAP_ENDPOINT_V41, ns='http://soa
-            cookie = responseInfo.getheader('set-cookie')
+            cookie = responseInfo.get('set-cookie')  # Python 3: getheader() -> get()
+            
+            # Debug: Show ALL Set-Cookie headers
```
Security concern as per above.

edalex-ian commented on source/equellaclient41.py:
```diff
@@ -384,11 +420,18 @@ def _createSoapSessionFromToken (self, token):
+        print(f"Login successful!")
+        print(f"Session cookies received: {len(self._cookieJar)} cookie(s)")
+        if len(self._cookieJar) > 0:
+            print(f"Cookie details: {self._cookieJar[:2]}")
```
Security concern.

edalex-ian commented on source/equellaclient41.py:
```diff
@@ -267,27 +293,38 @@ def _call (self, name, args, returns=1, facade=SOAP_ENDPOINT_V41, ns='http://soa
+                                if not name in ["PATH", "DOMAIN", "EXPIRES", "SECURE", "HTTPONLY"] and not name in added_cookie:
+                                    # save cookie
+                                    self._cookieJar.append(cookie_name)
+                                    added_cookie.append(name)
```
This deep nesting etc makes for very brittle hard to maintain code. It should be extracted into a function such as:

```python
def _process_response_cookies(self, response_info, known_cookie_names):
    """
    Extracts and stores valid cookies from response headers.
    Filters out reserved keywords and duplicates.
    """
    RESERVED_KEYS = {"PATH", "DOMAIN", "EXPIRES", "SECURE", "HTTPONLY"}
    
    # Declaratively flatten headers -> parts -> stripped strings
    headers = response_info.get_all('set-cookie') or []
    cookie_parts = (
        part.strip() 
        for header in headers if header 
        for part in header.split(',')
    )

    for cookie in cookie_parts:
        if not cookie: continue
            
        # Isolate name extraction logic
        key_segment = cookie.split("=")[0].strip().upper()
        name = key_segment.split(";")[1].strip() if ";" in key_segment else key_segment

        # Filter negative conditionals by checking membership
        if name not in RESERVED_KEYS and name not in known_cookie_names:
            self._cookieJar.append(cookie)
            known_cookie_names.append(name)
```
And then usage in main _call function:

```python
self._process_response_cookies(responseInfo, added_cookie)
```

edalex-ian commented on source/equellaclient41.py:
```diff
@@ -822,81 +893,154 @@ def read_in_chunks(self, file_object, chunk_size=(1024 * 1024)):
-    # Parent directories are automatically created as required.
-    # Uploads file in chunks of 16MB. If file is large (e.g. over 16MB) pass in parameter showstatus as a prefix to a progress report
-    # e.g. item.attachFile ('support/song.wav', file ('video.avi', 'rb'), '     Progress: ')
     def attachFile (self, path, attachment, showstatus=None, chunk_size=(1024 * 2048)):
```
The refactor of this file swapped things around a little, but possibly made it no worse than it was. However for new eyes coming in it's painful to read.

The current function struggles with **Single Responsibility Principle** violations; it attempts to validate inputs, manage complex UI side effects (logging, stdout, GUI updates), and handle binary file transmission simultaneously. This results in deep nesting (Arrow Code) and "Opacity," making the flow hard to scan.

To improve **Readability** and **Maintainability**, we should extract the "noisy" UI progress reporting into a dedicated method and separate the file validation. This allows the main function to tell a linear story: validate, calculate size, and iterate through chunks to upload.

### Refactored Example

I have separated the side effects (UI updates) from the business logic and flattened the control flow.

```python
import os
import sys
from binascii import b2a_base64

# Defined constant to replace magic number
DEFAULT_CHUNK_SIZE = 1024 * 2048 

def attach_file(self, path, attachment, show_status=None, chunk_size=DEFAULT_CHUNK_SIZE):
    """
    Orchestrates the file upload process. 
    Focuses on flow control rather than implementation details.
    """
    self._validate_attachment(attachment) # Fail fast
    
    try:
        file_size = os.path.getsize(attachment.name)
    except OSError as e:
        raise Exception(f"Cannot access file '{attachment.name}': {e}")

    uploaded_bytes = 0
    self._notify_progress(show_status, "Uploading...", 0, file_size)

    try:
        # enumerate gives us an index to determine first_chunk cleanly
        for i, chunk in enumerate(self.read_in_chunks(attachment, chunk_size)):
            # 1. Handle User Interruption
            wx.GetApp().Yield()
            if self.owner.StopProcessing:
                self._notify_halt(show_status)
                break

            # 2. Upload Logic
            encoded_chunk = b2a_base64(chunk).decode('ascii').strip()
            is_first_chunk = (i == 0)
            self.parClient._uploadFile(self.stagingid, path, encoded_chunk, is_first_chunk)

            # 3. Update State
            uploaded_bytes += len(chunk)
            self._notify_progress(show_status, "Uploading...", uploaded_bytes, file_size)
            
    except Exception:
        # Ensure newline on error if writing to stdout, then re-raise
        if not self.debug and sys.stdout: 
            print() 
        raise

def _validate_attachment(self, attachment):
    """Encapsulates boundary conditions"""
    if attachment is None:
        raise Exception("Attachment file object is None")
    if not hasattr(attachment, 'read') or not hasattr(attachment, 'name'):
        raise Exception("Attachment is not a valid file object")

def _notify_progress(self, prefix, state, current, total):
    """
    Isolates UI/Logging side effects from business logic.
   
    """
    if not prefix:
        return

    percent = (current * 100) // total if total > 0 else 0
    message = f"{prefix} {state}"
    
    if self.debug:
        # Simplified debug reporting
        report = f"{message} chunk uploaded. {current}/{total} ({percent}%)"
        self.owner.echo(report)
        if current >= total:
            self.owner.echo("    Done")
        self.owner.tryPausing("    [Paused]")
    else:
        # Production logging/stdout reporting
        status_line = f"{message}{percent}%" if current < total else f"{message}Done"
        self._write_to_logs(status_line)
        self._write_to_stdout("." if current < total else "Done\n")
        self.owner.tryPausing("    [Paused]", newline=True)

def _notify_halt(self, prefix):
    """Specific handler for user cancellation."""
    msg = f"{prefix} Halted by user" if prefix else "Halted by user"
    if self.debug:
        self.owner.echo(msg)
    else:
        self._write_to_logs("\n")
        self._write_to_stdout("Halted by user\n")
        self.owner.echo(msg, False)

def _write_to_logs(self, text):
    """Helper to handle the verbose log object locking/unlocking."""
    if self.owner.log:
        self.owner.log.SetReadOnly(False)
        # If it's a progress update, replace the line, otherwise append
        if "Uploading..." in text and not text.endswith("Done"):
             self.owner.log.DelLineLeft()
        self.owner.log.AppendText(text)
        self.owner.log.SetReadOnly(True)

def _write_to_stdout(self, text):
    """Safe stdout wrapper."""
    if sys.stdout:
        try:
            sys.stdout.write(text)
            sys.stdout.flush()
        except Exception:
            pass

```

edalex-ian commented on source/equellaclient41.py:
```diff
@@ -978,28 +1142,88 @@ def addStartPage (self, description, path, size=1024, uuid='', thumbnail = ""):
         attachment.createNode("description", description)
         if uuid != '':
             attachment.createNode("uuid", uuid)
+    def addStartPageWithZipAttribute(self, description, path, size=1024, uuid='', zipParentUUID=''):
```
Missing whitespace.

```suggestion

    def addStartPageWithZipAttribute(self, description, path, size=1024, uuid='', zipParentUUID=''):
```

edalex-ian reviewed (changes requested):
I’ve completed an initial pass of this migration. Given the scale of the changes (+2,370/-932) and the complexity of the underlying legacy structures (e.g., functions exceeding 1,000 lines), a comprehensive line-by-line validation was unfortunately not feasible. While I have verified the core logic where possible, the high degree of **nesting** and **opacity** makes it difficult to guarantee full coverage.

I did also build and run (well, launch but not any bulk functions) locally. And that's further than I'd got with the previous code.

To improve maintainability moving forward, I suggest we:

* **Refactor via AI:** Use LLMs to break down these 'God functions' into **small**, **single-purpose** methods with **descriptive names**.
* **Automate Builds:** Implement a GitHub Actions workflow to ensure builds are verified across all platforms automatically.

I’ve left specific notes on the most critical areas, but I recommend we prioritize cleaning these 'campgrounds' in subsequent PRs to follow the **Boy Scout Rule**.

Commit:
Refactor source structure and extract utilities

- Extract common constants to [source/Constants.py](cci:7://file:///Users/nelson/gitlab/equella/openEQUELLA-ebi/source/Constants.py:0:0-0:0).
- Extract utility functions to [source/Utils.py](cci:7://file:///Users/nelson/gitlab/equella/openEQUELLA-ebi/source/Utils.py:0:0-0:0).
- Consolidate source code by removing `package-mac/source/ebi` and standardizing on `source/`.
- Refactor `equellaclient41.py` for better progress reporting, logging, and error handling.
- Update `MainFrame.py`, `Engine.py`, and [ebi.py](cci:7://file:///Users/nelson/gitlab/equella/openEQUELLA-ebi/source/ebi.py:0:0-0:0) to use the new `Constants` and `Utils` modules.
- Retain core application constants (Version, Copyright, License) in `ebi.py` as requested.

Commit:
Refactor RowProcessor, upgrade to Python 3, add CI/CD and OS-specific packaging

Commit:
refactor: Addressed issues raised by AI reviewer.
- Encapsulate state and data cleaning logic in `RowProcessor.py`
- Reduce network logging verbosity
- Refine receipt file generation conditions.

edalex-ian commented:
> Oh yes, I just clicked the "re-review request". :) Thanks @edalex-ian .

Sorry, I was a bit slow on noticing the email. I'll try and get in and re-review next week. :crossed_fingers:
